File: app/src/main/java/com/jobtick/activities/SavedTaskActivity.java
package com.jobtick.activities;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.ImageView;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.airbnb.lottie.LottieAnimationView;
import com.android.volley.AuthFailureError;
import com.android.volley.DefaultRetryPolicy;
import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.StringRequest;
import com.android.volley.toolbox.Volley;
import com.jobtick.R;
import com.jobtick.adapers.TaskListAdapter;
import com.jobtick.interfaces.OnRemoveSavedTaskListener;
import com.jobtick.models.TaskModel;
import com.jobtick.utils.Constant;
import com.jobtick.utils.ConstantKey;
import com.jobtick.utils.SessionManager;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

import butterknife.BindView;
import butterknife.ButterKnife;
import pl.droidsonroids.gif.GifImageView;

import static com.jobtick.pagination.PaginationListener.PAGE_START;

public class SavedTaskActivity extends ActivityBase implements TaskListAdapter.OnItemClickListener, SwipeRefreshLayout.OnRefreshListener, OnRemoveSavedTaskListener {


    private TaskListAdapter taskListAdapter;
    private int currentPage = PAGE_START;
    private boolean isLastPage = false;
    private int totalPage = 10;
    private boolean isLoading = false;

    private SessionManager sessionManager;

    @BindView(R.id.recyclerview_savedTask)
    RecyclerView recyclerViewStatus;

    @BindView(R.id.swipeRefresh)
    SwipeRefreshLayout swipeRefresh;

  //  @BindView(R.id.ivNoPosst)
    //public GifImageView ivNoPost;

    private LottieAnimationView lottieAnim;
/*
    @BindView(R.id.toolbar)
    MaterialToolbar toolbar;
*/

    @BindView(R.id.ivBack)
    ImageView ivBack;

    public static OnRemoveSavedTaskListener onRemoveSavedtasklistener;


    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_saved_task);
        ButterKnife.bind(this);

        lottieAnim=findViewById(R.id.lottieAnimationView);
        onRemoveSavedtasklistener = this;
        sessionManager = new SessionManager(SavedTaskActivity.this);
        LinearLayoutManager layoutManager = new LinearLayoutManager(SavedTaskActivity.this);
        recyclerViewStatus.setLayoutManager(layoutManager);
        taskListAdapter = new TaskListAdapter(new ArrayList<>());
        recyclerViewStatus.setAdapter(taskListAdapter);
        taskListAdapter.setOnItemClickListener(this);
        swipeRefresh.setOnRefreshListener(this);

        // swipeRefresh.setRefreshing(true);
        getStatusList();

        //toolbar.setNavigationOnClickListener(view -> onBackPressed());
        ivBack.setOnClickListener(v -> finish());

    }


    private void getStatusList() {


        swipeRefresh.setRefreshing(true);
        ArrayList<TaskModel> items = new ArrayList<>();
        StringRequest stringRequest = new StringRequest(Request.Method.GET, Constant.URL_BOOKMARK + "?page=" + currentPage,
                response -> {
                    swipeRefresh.setRefreshing(false);

                    Log.e("responce_url", response);
                    // categoryArrayList.clear();
                    try {
                        JSONObject jsonObject = new JSONObject(response);
                        Log.e("json", jsonObject.toString());
                        if (jsonObject.has("data") && !jsonObject.isNull("data")) {
                            JSONArray jsonArray_data = jsonObject.getJSONArray("data");
                            for (int i = 0; jsonArray_data.length() > i; i++) {
                                JSONObject jsonObject_taskModel_list = jsonArray_data.getJSONObject(i);
                                TaskModel taskModel = new TaskModel().getJsonToModel(jsonObject_taskModel_list.getJSONObject("task"), SavedTaskActivity.this);
                                items.add(taskModel);
                            }
                        } else {
                            showToast("some went to wrong", SavedTaskActivity.this);
                            return;
                        }

                        if (jsonObject.has("meta") && !jsonObject.isNull("meta")) {
                            JSONObject jsonObject_meta = jsonObject.getJSONObject("meta");
                            totalPage = jsonObject_meta.getInt("last_page");
                            Constant.PAGE_SIZE = jsonObject_meta.getInt("per_page");
                        }

                        /*
                         *manage progress view
                         */
                        if (currentPage != PAGE_START)
                            taskListAdapter.removeLoading();

                        if (items.size() <= 0) {
                            lottieAnim.setVisibility(View.VISIBLE);
                            recyclerViewStatus.setVisibility(View.GONE);
                        } else {
                            lottieAnim.setVisibility(View.GONE);
                            recyclerViewStatus.setVisibility(View.VISIBLE);

                        }
                        taskListAdapter.addItems(items);

                        swipeRefresh.setRefreshing(false);
                        // check weather is last page or not
                        if (currentPage < totalPage) {
                            taskListAdapter.addLoading();
                        } else {
                            isLastPage = true;
                        }
                        isLoading = false;
                    } catch (JSONException e) {
                        hideProgressDialog();
                        Log.e("EXCEPTION", String.valueOf(e));
                        e.printStackTrace();
                    }
                },
                new Response.ErrorListener() {
                    @Override
                    public void onErrorResponse(VolleyError error) {
                        //  swipeRefresh.setRefreshing(false);

                        swipeRefresh.setRefreshing(false);
                        errorHandle1(error.networkResponse);
                    }
                }) {
            @Override
            public Map<String, String> getHeaders() throws AuthFailureError {
                Map<String, String> map1 = new HashMap<String, String>();
                map1.put("Content-Type", "application/x-www-form-urlencoded");
                map1.put("Authorization", "Bearer " + sessionManager.getAccessToken());
                return map1;
            }
        };

        stringRequest.setRetryPolicy(new DefaultRetryPolicy(0, -1,
                DefaultRetryPolicy.DEFAULT_BACKOFF_MULT));
        RequestQueue requestQueue = Volley.newRequestQueue(SavedTaskActivity.this);
        requestQueue.add(stringRequest);
        Log.e("url", stringRequest.getUrl());

    }

    @Override
    public void onRefresh() {
        swipeRefresh.setRefreshing(true);
        currentPage = PAGE_START;
        isLastPage = false;
        taskListAdapter.clear();
        getStatusList();
    }

    @Override
    public void onItemClick(View view, TaskModel obj, int position, String action) {
        if (obj.getStatus().toLowerCase().equalsIgnoreCase(Constant.TASK_DRAFT.toLowerCase())) {
            //    getDataFromServer(obj.getSlug());
        } else {
            Intent intent = new Intent(SavedTaskActivity.this, TaskDetailsActivity.class);
            Bundle bundle = new Bundle();
            bundle.putString(ConstantKey.SLUG, obj.getSlug());
            bundle.putInt(ConstantKey.USER_ID, obj.getPoster().getId());
            intent.putExtras(bundle);
            startActivity(intent);
        }
    }

    @Override
    public void onRemoveSavedTask() {
        currentPage = PAGE_START;
        isLastPage = false;
        taskListAdapter.clear();
        getStatusList();
    }
}

File: app/src/main/java/com/jobtick/android/activities/SavedTaskActivity.kt
package com.jobtick.android.activities

import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.widget.LinearLayout
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout.OnRefreshListener
import com.android.volley.*
import com.android.volley.toolbox.StringRequest
import com.android.volley.toolbox.Volley
import com.google.android.material.appbar.MaterialToolbar
import com.jobtick.android.BuildConfig
import com.jobtick.android.R
import com.jobtick.android.adapers.TaskListAdapter
import com.jobtick.android.interfaces.OnRemoveSavedTaskListener
import com.jobtick.android.models.TaskModel
import com.jobtick.android.pagination.PaginationListener
import com.jobtick.android.utils.Constant
import com.jobtick.android.utils.ConstantKey
import com.jobtick.android.utils.SessionManager
import com.jobtick.android.viewmodel.SavedJobsViewModel
import org.json.JSONException
import org.json.JSONObject
import timber.log.Timber
import java.util.*

class SavedTaskActivity : ActivityBase(), TaskListAdapter.OnItemClickListener, OnRefreshListener,
    OnRemoveSavedTaskListener {
    private var taskListAdapter: TaskListAdapter? = null
    private var currentPage = PaginationListener.PAGE_START
    private var isLastPage = false
    private var totalPage = 10
    private var totalItem = 10
    private val isLoading = false
    private lateinit var sessionManagerL: SessionManager
    private lateinit var viewModel: SavedJobsViewModel
    private var recyclerViewStatus: RecyclerView? = null
    private var swipeRefresh: SwipeRefreshLayout? = null
    private var noPosts: LinearLayout? = null
    private var toolbar: MaterialToolbar? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_saved_task)
        setIDS()
        noPosts = findViewById(R.id.no_posts_container)
        onRemoveSavedtasklistener = this
        sessionManagerL = SessionManager(this@SavedTaskActivity)
        val layoutManager = LinearLayoutManager(this@SavedTaskActivity)
        recyclerViewStatus!!.layoutManager = layoutManager
        taskListAdapter = TaskListAdapter(ArrayList(), null)
        recyclerViewStatus!!.adapter = taskListAdapter
        taskListAdapter!!.setOnItemClickListener(this)
        swipeRefresh!!.setOnRefreshListener(this)
        initToolbar()

        initVM()
        // swipeRefresh.setRefreshing(true);
        getStatusList()
    }

    private fun initVM() {
        viewModel = ViewModelProvider(this).get(SavedJobsViewModel::class.java)
        val items = ArrayList<TaskModel>()

        viewModel.getJobsResponse().observe(this, androidx.lifecycle.Observer {
            swipeRefresh!!.isRefreshing = false
            // categoryArrayList.clear();
            try {
                val jsonObject = it
                Timber.e(jsonObject.toString())
                if (jsonObject.has("data") && !jsonObject.isNull("data")) {
                    val jsonArray_data = jsonObject.getJSONArray("data")
                    var i = 0
                    while (jsonArray_data.length() > i) {
                        val jsonObject_taskModel_list = jsonArray_data.getJSONObject(i)
                        val taskModel = TaskModel().getJsonToModel(
                            jsonObject_taskModel_list.getJSONObject("task"),
                            this@SavedTaskActivity
                        )
                        items.add(taskModel)
                        i++
                    }
                } else {
                    showToast("some went to wrong", this@SavedTaskActivity)
                    return@Observer
                }
                if (jsonObject.has("meta") && !jsonObject.isNull("meta")) {
                    val jsonObject_meta = jsonObject.getJSONObject("meta")
                    totalPage = jsonObject_meta.getInt("last_page")
                    totalItem = jsonObject_meta.getInt("total")
                    Constant.PAGE_SIZE = jsonObject_meta.getInt("per_page")
                }
                if (items.size <= 0) {
                    noPosts!!.visibility = View.VISIBLE
                    recyclerViewStatus!!.visibility = View.GONE
                } else {
                    noPosts!!.visibility = View.GONE
                    recyclerViewStatus!!.visibility = View.VISIBLE
                }
                taskListAdapter!!.addItems(items, totalItem)
                swipeRefresh!!.isRefreshing = false
            } catch (e: JSONException) {
                hideProgressDialog()
                Timber.e(e.toString())
                e.printStackTrace()
            }
        })
    }

    private fun setIDS() {
        recyclerViewStatus = findViewById(R.id.recyclerview_savedTask)
        swipeRefresh = findViewById(R.id.swipeRefresh)
        toolbar = findViewById(R.id.toolbar)
    }

    private fun initToolbar() {
        toolbar!!.setNavigationIcon(R.drawable.ic_back)
        setSupportActionBar(toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        supportActionBar!!.title = "Saved Jobs"
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            onBackPressed()
        }
        return super.onOptionsItemSelected(item)
    }//  swipeRefresh.setRefreshing(false);

    // categoryArrayList.clear();
    private fun getStatusList() {
        swipeRefresh!!.isRefreshing = true
        viewModel.getJobs(
            sessionManager!!.accessToken,
            Volley.newRequestQueue(this),
            currentPage.toString()
        )
    }

    override fun onRefresh() {
        swipeRefresh!!.isRefreshing = true
        currentPage = PaginationListener.PAGE_START
        isLastPage = false
        taskListAdapter!!.clear()
        getStatusList()
    }

    override fun onItemClick(view: View, obj: TaskModel, position: Int, action: String) {
        if (obj.status.toLowerCase().equals(Constant.TASK_DRAFT.toLowerCase(), ignoreCase = true)) {
            //    getDataFromServer(obj.getSlug());
        } else {
            val intent = Intent(this@SavedTaskActivity, TaskDetailsActivity::class.java)
            val bundle = Bundle()
            bundle.putString(ConstantKey.SLUG, obj.slug)
            //   bundle.putInt(ConstantKey.USER_ID, obj.getPoster().getId());
            intent.putExtras(bundle)
            startActivity(intent)
        }
    }

    override fun onRemoveSavedTask() {
        currentPage = PaginationListener.PAGE_START
        isLastPage = false
        taskListAdapter!!.clear()
        getStatusList()
    }

    companion object {
        @JvmField
        var onRemoveSavedtasklistener: OnRemoveSavedTaskListener? = null
    }
}